Submitted By: Robert Connolly <cendres at videotron dot ca> (ashes)
Date: 2003-02-15
Initial Package Version: 2.6.2
Origin: http://www.northernsecurity.net/adamantix/diffs/2.6.0-propolice.patch
Description: Linux kernel patch for Smashing Stack Protector. This patch allows
the use of -fstack-protector durring the kernel build.
https://twocents.mooo.com/hints/downloads/files/winter.txt
https://twocents.mooo.com/hints/downloads/files/ssp.txt

diff -Naur linux-2.6.2.orig/Makefile linux-2.6.2/Makefile
--- linux-2.6.2.orig/Makefile	2004-02-04 03:44:04.000000000 +0000
+++ linux-2.6.2/Makefile	2004-02-16 01:15:45.000000000 +0000
@@ -277,7 +277,7 @@
 		   $(if $(KBUILD_SRC),-Iinclude2 -I$(srctree)/include)
 
 CFLAGS 		:= -Wall -Wstrict-prototypes -Wno-trigraphs \
-	  	   -fno-strict-aliasing -fno-common
+	  	   -fno-strict-aliasing -fno-common -fstack-protector
 AFLAGS		:= -D__ASSEMBLY__
 
 export	VERSION PATCHLEVEL SUBLEVEL EXTRAVERSION KERNELRELEASE ARCH \
diff -Naur linux-2.6.2.orig/arch/i386/kernel/i386_ksyms.c linux-2.6.2/arch/i386/kernel/i386_ksyms.c
--- linux-2.6.2.orig/arch/i386/kernel/i386_ksyms.c	2004-02-04 03:45:04.000000000 +0000
+++ linux-2.6.2/arch/i386/kernel/i386_ksyms.c	2004-02-16 01:15:45.000000000 +0000
@@ -97,6 +97,11 @@
 EXPORT_SYMBOL_NOVERS(__down_failed_interruptible);
 EXPORT_SYMBOL_NOVERS(__down_failed_trylock);
 EXPORT_SYMBOL_NOVERS(__up_wakeup);
+
+/* SSP */
+EXPORT_SYMBOL_NOVERS(__guard);
+EXPORT_SYMBOL_NOVERS(__stack_smash_handler);
+
 /* Networking helper routines. */
 EXPORT_SYMBOL(csum_partial_copy_generic);
 /* Delay loops */
diff -Naur linux-2.6.2.orig/arch/um/kernel/ksyms.c linux-2.6.2/arch/um/kernel/ksyms.c
--- linux-2.6.2.orig/arch/um/kernel/ksyms.c	2004-02-04 03:43:07.000000000 +0000
+++ linux-2.6.2/arch/um/kernel/ksyms.c	2004-02-16 01:15:45.000000000 +0000
@@ -90,3 +90,5 @@
 EXPORT_SYMBOL(kmap_atomic_to_page);
 #endif
 
+EXPORT_SYMBOL_NOVERS(__guard);
+EXPORT_SYMBOL_NOVERS(__stack_smash_handler);
diff -Naur linux-2.6.2.orig/include/linux/kernel.h linux-2.6.2/include/linux/kernel.h
--- linux-2.6.2.orig/include/linux/kernel.h	2004-02-04 03:43:04.000000000 +0000
+++ linux-2.6.2/include/linux/kernel.h	2004-02-16 01:15:45.000000000 +0000
@@ -114,6 +114,8 @@
 #define TAINT_FORCED_RMMOD		(1<<3)
 
 extern void dump_stack(void);
+extern int __guard;
+extern void __stack_smash_handler(int, char []);
 
 #ifdef DEBUG
 #define pr_debug(fmt,arg...) \
diff -Naur linux-2.6.2.orig/lib/Makefile linux-2.6.2/lib/Makefile
--- linux-2.6.2.orig/lib/Makefile	2004-02-04 03:43:07.000000000 +0000
+++ linux-2.6.2/lib/Makefile	2004-02-16 01:16:21.000000000 +0000
@@ -6,7 +6,7 @@
 lib-y := errno.o ctype.o string.o vsprintf.o cmdline.o \
 	 bust_spinlocks.o rbtree.o radix-tree.o dump_stack.o \
 	 kobject.o idr.o div64.o parser.o int_sqrt.o mask.o \
-	 bitmap.o extable.o
+	 bitmap.o extable.o propolice.o
 
 lib-$(CONFIG_RWSEM_GENERIC_SPINLOCK) += rwsem-spinlock.o
 lib-$(CONFIG_RWSEM_XCHGADD_ALGORITHM) += rwsem.o
diff -Naur linux-2.6.2.orig/lib/propolice.c linux-2.6.2/lib/propolice.c
--- linux-2.6.2.orig/lib/propolice.c	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.2/lib/propolice.c	2004-02-16 01:15:45.000000000 +0000
@@ -0,0 +1,13 @@
+ /*
+  *  linux/lib/errno.c
+  *
+  *
+  */
+ 
+ int __guard = '\0\0\n\777';
+ 
+ void __stack_smash_handler (int damaged, char func[])
+ 	 {
+	   static char *message = "propolice detects %x at function %s.\n" ;
+	   panic (message, damaged, func);
+}
